engine/baml-runtime/src/eval_expr.rs

```diff
@@ -0,0 +1,488 @@
+                return Ok(Some(BamlValueWithMeta::Map(new_items, ())));
+            }
+            Expr::ClassConstructor {
+                name,
+                fields,
+                spread,
+                meta,
+            } => {
+                let stream: Vec<Result<(String, BamlValueWithMeta<()>), anyhow::Error>> =
+                    stream::iter(fields.iter().map(|(key, value)| async move {
+                        let res = Box::pin(eval_to_value(env, value))
+                            .await?
+                            .ok_or(anyhow::anyhow!("Failed to evaluate class constructor"))?;
+                        Ok((key.clone(), res))
+                    }))
+                    .buffered(2)
+                    .collect::<Vec<Result<(String, BamlValueWithMeta<()>), _>>>()
+                    .await;
+                let fields = stream
+                    .into_iter()
+                    .collect::<Result<Vec<(String, BamlValueWithMeta<()>)>, anyhow::Error>>()?;
+
+                let mut spread_fields = match spread {
+                    Some(spread) => {
+                        let res = Box::pin(eval_to_value(env, spread.as_ref())).await?;
+                        match res {
+                            Some(BamlValueWithMeta::Class(spread_class_name, spread_fields, _)) => {
+                                if name != spread_class_name {
+                                    return Err(anyhow::anyhow!("Class constructor name mismatch"));
+                                }
+                                spread_fields.clone()
+                            }
+                            _ => {
+                                return Err(anyhow::anyhow!("Spread is not a class"));
+                            }
+                        }
+                    }
+                    None => BamlMap::new(),
+                };
+
+                spread_fields.extend(fields);
+                let val = BamlValueWithMeta::Class(name.clone(), spread_fields, ());
+                return Ok(Some(val));
+            }
+            other => {
+                // let new_expr = step(env, &other).await?;
+                let new_expr = Box::pin(beta_reduce(env, &other)).await?;
+
+                if new_expr.temporary_same_state(expr) {
+                    return Err(anyhow::anyhow!("Failed to make progress."));
+                }
+                current_expr = new_expr;
+            }
+        }
+    }
+    Err(anyhow::anyhow!("Max steps reached."))
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::internal_baml_diagnostics::Span;
+    use baml_types::{BamlMap, BamlValue};
+    use futures::channel::mpsc;
+    use internal_baml_core::ir::repr::make_test_ir;
+
+    use super::*;
+    use crate::BamlRuntime;
+
+    // Make a testing runtime. It assumes the presence of
+    // OPENAI_API_KEY environment variable.
+    fn runtime(content: &str) -> BamlRuntime {
```

Comment:
does this mean `cargo test` will now fail by default?

Reply:
If any unit tests actually uses this function, yes. Right now it's just a convenience thing to use during development. I tweak the inline BAML code starting on line 386, then uncomment line 382 to enable the test.

Probably should come up with something better. Integ test would make sense, and I think we'd be able to do that as soon as we wire expr functions up to codegen.

Reply:
my question the other day at standup around `cargo test` only supporting pass/fail and not more complex statuses was basically about this

a bunch of the `baml-cli serve` tests have this same issue right now

Reply:
Heh, agreed!